xayn-ai/src/coi/system.rs

```diff
@@ -214,6 +223,160 @@ impl CoiSystem {
             })
             .collect()
     }
+
+    /// Selects the most relevant key phrases for the coi.
+    ///
+    /// The most relevant key phrases are selected from the set of key phrases of the coi and the
+    /// candidates. The computed relevances are a relative score from the interval `[0, 1]`.
+    #[allow(dead_code)]
+    fn select_key_phrases<
+        CP: CoiPoint + CoiPointKeyPhrases,
+        F: Fn(&str) -> Result<Embedding, Error>,
+    >(
+        &self,
+        coi: &mut CP,
+        candidates: &[String],
+        // TODO: make SMBert available to CoiSystem and remove this argument
+        smbert: F,
+    ) {
+        /// Reduces the matrix along the axis while skipping the diagonal elements.
+        fn reduce_without_diag<S, F, G>(a: ArrayBase<S, Ix2>, axis: Axis, f: F, g: G) -> Array2<f32>
+        where
+            S: Data<Elem = f32>,
+            F: Fn(f32, f32) -> f32,
+            G: Fn(f32) -> f32,
+        {
+            a.lanes(axis)
+                .into_iter()
+                .enumerate()
+                .map(|(i, lane)| {
+                    lane.iter()
+                        .enumerate()
+                        .filter_map(|(j, element)| (i != j).then(|| *element))
+                        .reduce(|reduced, element| f(reduced, element))
+                        .map(|reduced| g(reduced))
```

Comment:
Locally clippy suggests this:
```suggestion
                        .reduce(f)
                        .map(g)
```

Reply:
that's odd, clippy doesn't say so for me. if i replace it as suggested, then i get errros like 

```
cannot move out of `f`, a captured variable in an `FnMut` closure
move occurs because `f` has type `F`, which does not implement the `Copy` trait
```

i'll add the `Copy` bound to `F` and `G`, this makes it less general, but it's a private helper function anyways so it shouldn't matter much.

Reply:
I had 1.57 and changing with the suggested code it throw the same error :/ :(